cdap-common/bin/functions.sh

```diff
@@ -0,0 +1,1085 @@
+  if [[ -z ${__principal} ]] || [[ -z ${__keytab} ]]; then
+    die "Both cdap.master.kerberos.principal and cdap.master.kerberos.keytab must be configured for Kerberos"
+  fi
+  if [[ ! -r ${__keytab} ]]; then
+    die "Cannot read keytab: ${__keytab}"
+  fi
+  if [[ $(which kinit 2>/dev/null) ]]; then
+    # Replace _HOST in principal w/ FQDN, like Hadoop does
+    kinit -kt "${__keytab}" "${__principal/_HOST/${HOSTNAME}}"
+    if [[ ! ${?} ]]; then
+      die "Failed executing 'kinit -kt \"${__keytab}\" \"${__principal/_HOST/${HOSTNAME}}\"'"
+    fi
+  else
+    die "Cannot locate kinit! Please, ensure the appropriate Kerberos utilities are installed"
+  fi
+  return 0
+}
+
+#
+# cdap_set_java
+# Attempts to find JAVA in few ways and sets JAVA variable
+#
+cdap_set_java () {
+  local readonly __java __java_version
+  # Check JAVA_HOME, first
+  if [[ -n ${JAVA_HOME} ]] && [[ -d ${JAVA_HOME} ]]; then
+    __java="${JAVA_HOME}"/bin/java
+    [[ -x ${__java} ]] || die "JAVA_HOME is set to an invalid location: ${JAVA_HOME}"
+  else
+    __java=${JAVA:-java}
+    if [[ $(which java 2>/dev/null) ]]; then
+      : # continue
+    else
+      die "JAVA_HOME is not set and 'java' was not found in your PATH. Please set JAVA_HOME to the location of your Java installation"
+    fi
+  fi
+  __java_version=$("${__java}" -version 2>&1 | grep version | awk '{print $3}' | awk -F '.' '{print $2}')
+  if [[ -z ${__java_version} ]]; then
+    die "Could not detect Java version. Aborting..."
+  elif [[ ${__java_version} -ne 7 ]] && [[ ${__java_version} -ne 8 ]]; then
+    die "Java version not supported. Please install Java 7 or 8 - other versions of Java are not supported."
+  fi
+  export JAVA=${__java}
+  return 0
+}
+
+#
+# cdap_set_classpath <home-dir> <conf-dir> [verbose: true/false]
+# Assembles CLASSPATH from home-dir, hbase classpath, and conf-dir and optionally echos if verbose is set true
+# NOTE: this function is also sourced and invoked by the CSD control script, found here:
+#   https://github.com/caskdata/cm_csd/blob/develop/src/scripts/cdap-control.sh
+#   Any changes to this function must be compatible with the CSD's invocation
+#
+cdap_set_classpath() {
+  local readonly __home=${1} __conf=${2} __verbose=${3:-false}
+  local readonly __homelib=$(find -L "${__home}"/lib -type f 2>/dev/null | sort | tr '\n' ':')
+  local __cp __hbase_cp
+
+  # Get HBase's CLASSPATH
+  if [[ -n ${HBASE_CLASSPATH} ]] && [[ ${HBASE_CLASSPATH} != '' ]]; then
+    __cp=${__homelib}:${HBASE_CLASSPATH}:${__conf}/:${__home}/conf/:${EXTRA_CLASSPATH}
+  elif [[ -n ${HBASE_HOME} ]] && [[ -d ${HBASE_HOME} ]]; then
+    __hbase_cp=$("${HBASE_HOME}"/bin/hbase classpath)
+  elif [[ $(which hbase 2>/dev/null) ]]; then
+    __hbase_cp=$(hbase classpath)
+  else
+    # assume Hadoop/HBase libs are included via EXTRA_CLASSPATH
+    echo "[WARN] Could not find Hadoop and HBase libraries, using EXTRA_CLASSPATH"
```

Comment:
this and several other places should probably use ``logecho`` (preexisting issue, not a blocker)